contracts/protocol/v1/proxies/P1LiquidatorProxy.sol

```diff
@@ -0,0 +1,288 @@
+    }
+
+    // ============ Helper Functions ============
+
+    /**
+     * @dev Calculate (and verify) the maximum amount to liquidate based on the maxPosition input.
+     */
+    function _getMaxPositionDelta(
+        P1Types.Balance memory initialBalance,
+        bool isBuy,
+        SignedMath.Int memory maxPosition
+    )
+        private
+        pure
+        returns (SignedMath.Int memory)
+    {
+        SignedMath.Int memory result = maxPosition.signedSub(initialBalance.getPosition());
+
+        require(
+            result.isPositive == isBuy && result.value > 0,
+            "Cannot liquidate past maxPosition"
+        );
+
+        return result;
+    }
+
+    /**
+     * @dev Perform the liquidation by constructing the correct arguments and sending it to the
+     * perpetual.
+     */
+    function _doLiquidation(
+        I_PerpetualV1 perpetual,
+        address liquidatee,
+        SignedMath.Int memory maxPositionDelta
+    )
+        private
+    {
+        // Create accounts.
+        bool takerFirst = address(msg.sender) < liquidatee;
+        address[] memory accounts = new address[](2);
+        uint256 takerIndex = takerFirst ? 0 : 1;
+        uint256 makerIndex = takerFirst ? 1 : 0;
+        accounts[takerIndex] = msg.sender;
+        accounts[makerIndex] = liquidatee;
+
+        // Create trade args.
+        I_PerpetualV1.TradeArg[] memory trades = new I_PerpetualV1.TradeArg[](1);
+        trades[0] = I_PerpetualV1.TradeArg({
+            takerIndex: takerIndex,
+            makerIndex: makerIndex,
+            trader: _LIQUIDATION_,
+            data: abi.encode(
+                maxPositionDelta.value,
+                maxPositionDelta.isPositive,
+                false // allOrNothing
+            )
+        });
+
+        // Do the liquidation.
+        perpetual.trade(accounts, trades);
+    }
+
+    /**
+     * @dev Calculate the liquidated amount and also the fee amount based on a percentage of the
+     * value of the repaid debt.
+     * @return  The position amount bought or sold.
+     * @return  The fee amount in margin token.
+     */
+    function _getLiquidatedAndFeeAmount(
+        I_PerpetualV1 perpetual,
+        bool isBuy,
+        P1Types.Balance memory initialBalance,
+        P1Types.Balance memory currentBalance
+    )
+        private
+        view
+        returns (uint256, uint256)
+    {
+        // Get the amount liquidated.
+        uint256 liqAmount = currentBalance.getPosition().signedSub(
+            initialBalance.getPosition()
+        ).value;
+
+        // Get the value of debt liquidated.
+        uint256 debtAmountInMargin = isBuy
+            ? currentBalance.getMargin().signedSub(initialBalance.getMargin()).value
+            : liqAmount.baseMul(perpetual.getOraclePrice());
+
+        // Calculate the fee (in margin tokens) based on the value of the debt liquidated.
+        uint256 feeAmount = debtAmountInMargin.baseMul(_FEE_PERCENTAGE_);
```

Comment:
If you want `_FEE_PERCENTAGE_` to be a percentage, shouldn't you divide by 100 here?

Reply:
Will change the name to be more clear